### DISCUSSION TITLE
Stamp with engine

### DISCUSSION THREAD

**Discussion by hyperknot** [sqlalchemy/alembic#888] [Usage Questions]
I think mine is a super common use case, setting up db + stamping, so much that it's the #1 item in the cookbook:
https://alembic.sqlalchemy.org/en/latest/cookbook.html

```
my_metadata.create_all(engine)
alembic_cfg = Config("/path/to/yourapp/alembic.ini")
command.stamp(alembic_cfg, "head")
```

I'd like to stamp using an engine. I believe it's typical to set up your DB server, extensions, views, create_all(), etc. all using an engine. Afterwards you'd need to stamp head.

Right now I cannot do this, as I'm setting up the engine programmatically and stamping requires an ini file.

**Comment by hyperknot** [comment]
Created issue https://github.com/sqlalchemy/alembic/issues/889

**Comment by CaselIT** [comment]
Well alembic needs a way of locating the version directory ect, so the ini file is still needed

**Reply by hyperknot** [reply]
That could be a simple argument to stamp, like it's done in an other place: `def stamp(self, script_directory, revision)`.

**Reply by zzzeek** [reply]
what database URL would stamp() connect to ?

**Comment by zzzeek** [comment]
running commands with specific engines is illustrated at https://alembic.sqlalchemy.org/en/latest/cookbook.html#sharing-a-connection-with-a-series-of-migration-commands-and-environments

```py
my_metadata.create_all(engine)
alembic_cfg = Config()
alembic_cfg.attributes['engine'] = engine
command.stamp(alembic_cfg, "head")
```

place this in your env.py:

```py
def run_migrations_online():
    engine = config.attributes.get("engine", None)

    if engine is None:
        engine = engine_from_config(
            config.get_section(config.config_ini_section),
            prefix='sqlalchemy.',
            poolclass=pool.NullPool)
...
```

**Reply by hyperknot** [reply]
Where do I specify the script_directory in this case?

**Reply by zzzeek** [reply]
if you dont want to use an alembic.ini file you can set it up manually in the Config(), see https://alembic.sqlalchemy.org/en/latest/api/config.html#alembic.config.Config

```py
from alembic.config import Config
alembic_cfg = Config()
alembic_cfg.set_main_option("script_location", "myapp:migrations")
alembic_cfg.set_main_option("sqlalchemy.url", "postgresql://foo/bar")
alembic_cfg.set_section_option("mysection", "foo", "bar")
```

**Reply by hyperknot** [reply]
So for a stamp(), specifying only the following 2 would work?

```py
alembic_cfg = Config()
alembic_cfg.attributes['engine'] = engine
alembic_cfg.set_main_option("script_location", "myapp:migrations")
command.stamp(alembic_cfg, "head")
```

**Reply by zzzeek** [reply]
probably, if it complains about other values we can address.     you need to change your env.py though to accommodate the "engine" attribute.
